pczt/src/orchard.rs

```diff
@@ -0,0 +1,428 @@
+    /// - The IO Finalizer compresses it into the bsk.
+    /// - This is required by the Prover.
+    /// - This may be used by Signers to verify that the value correctly matches `cv`.
+    ///
+    /// This opens `cv` for all participants. For Signers who don't need this information,
+    /// or after proofs / signatures have been applied, this can be redacted.
+    pub(crate) rcv: Option<[u8; 32]>,
+}
+
+/// Information about a Sapling spend within a transaction.
+#[derive(Clone)]
+pub(crate) struct Spend {
+    //
+    // Spend-specific Action effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) nullifier: [u8; 32],
+    pub(crate) rk: [u8; 32],
+
+    /// The spend authorization signature.
+    ///
+    /// This is set by the Signer.
+    pub(crate) spend_auth_sig: Option<[u8; 64]>,
+
+    /// The address that received the note being spent.
+    ///
+    /// - This is set by the Constructor (or Updater?).
+    /// - This is required by the Prover.
+    pub(crate) recipient: Option<[u8; 43]>,
+
+    /// The value of the input being spent.
+    ///
+    /// - This is required by the Prover.
+    /// - This may be used by Signers to verify that the value matches `cv`, and to
+    ///   confirm the values and change involved in the transaction.
+    ///
+    /// This exposes the input value to all participants. For Signers who don't need this
+    /// information, or after signatures have been applied, this can be redacted.
+    pub(crate) value: Option<u64>,
+
+    /// The rho value for the note being spent.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    ///
+    /// TODO: This could be merged with `rseed` into a tuple. `recipient` and `value` are
+    /// separate because they might need to be independently redacted. (For which role?)
+    pub(crate) rho: Option<[u8; 32]>,
+
+    /// The seed randomness for the note being spent.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The full viewing key that received the note being spent.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    pub(crate) fvk: Option<[u8; 96]>,
+
+    /// A witness from the note to the bundle's anchor.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    pub(crate) witness: Option<(u32, [[u8; 32]; 32])>,
+
+    /// The spend authorization randomizer.
+    ///
+    /// - This is chosen by the Constructor.
+    /// - This is required by the Signer for creating `spend_auth_sig`, and may be used to
+    ///   validate `rk`.
+    /// - After`zkproof` / `spend_auth_sig` has been set, this can be redacted.
+    pub(crate) alpha: Option<[u8; 32]>,
+
+    // TODO derivation path
+
+    // TODO FROST
```

Comment:
For FROST I think we need:

- Maybe `SigningPackage` (commitments and message, in this case, the sighash which will need to be recalculated and checked by the receiver). We could do the other way around though; send the PCZT inside the message field of the `SigningPackage` and replace it with the sighash before proceeding. Or just send the two together (this is what I planned to do in the demo repo, in [a struct](https://github.com/ZcashFoundation/frost-zcash-demo/blob/3338161fc10c13a025c8e45e84903a6f3acd16da/server/src/types.rs#L107) that contains both the SigningPackage and the PCZT in the `aux_msg` field)
- Possibly `randomizer_seed` which is used to rederive `alpha` and which will be checked by the receiver, per current https://github.com/zcash/zips/pull/895. But similarly this could be done outside the PCZT.

I'm not sure we need to use PCZTs for the round 2 reply from Participants to Coordinator; it would be just a gigantic wrapper for a signature share where everything but the signature share would be thrown away anyway.

Reply:
> I'm not sure we need to use PCZTs for the round 2 reply from Participants to Coordinator; it would be just a gigantic wrapper for a signature share where everything but the signature share would be thrown away anyway.

You may not need to in some cases, but I can imagine workflows where the round is asynchronous and the coordinator needs to learn which signing session the signature share(s) belong to. PCZT gives you that for free: the Combiner merges the PCZT results from each participant into a single PCZT, and then the Spend Finalizer runs the FROST aggregation step.